types/validator_set.go

```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+		_, val := vals.GetByAddress(address)
+		if val == nil { // add
+			updatedTotalVotingPower += valUpdate.VotingPower
+		} else { // update
+			updatedTotalVotingPower += valUpdate.VotingPower - val.VotingPower
+		}
+
+		overflow := updatedTotalVotingPower > MaxTotalVotingPower || updatedTotalVotingPower < 0
+		if overflow {
+			err = fmt.Errorf(
+				"failed to add/update validator %v. total voting power would exceed the max allowed %v",
+				valUpdate, MaxTotalVotingPower)
+			return
+		}
+	}
+
+	return
+}
+
+// Computes the proposer priority for the validators not present in the set based on 'updatedTotalVotingPower'
+// Leaves unchanged the priorities of validators that are changed.
+//
+// 'updates' parameter must be a list of unique validators to be added or updated.
+// No changes are made to the validator set 'vals'.
+func computeNewPriorities(updates []*Validator, vals *ValidatorSet, updatedTotalVotingPower int64) {
+
+	// Scan and update the proposerPriority for newly added and updated validators
+	for _, valUpdate := range updates {
+		address := valUpdate.Address
+		_, val := vals.GetByAddress(address)
+		if val == nil {
+			// add val
+			// Set ProposerPriority to -C*totalVotingPower (with C ~= 1.125) to make sure validators can't
+			// un-bond and then re-bond to reset their (potentially previously negative) ProposerPriority to zero.
+			//
+			// Contract: updatedVotingPower < MaxTotalVotingPower to ensure ProposerPriority does
+			// not exceed the bounds of int64.
+			//
+			// Compute ProposerPriority = -1.125*totalVotingPower == -(updatedVotingPower + (updatedVotingPower >> 3)).
+			valUpdate.ProposerPriority = -(updatedTotalVotingPower + (updatedTotalVotingPower >> 3))
+		} else {
+			valUpdate.ProposerPriority = val.ProposerPriority
+		}
+	}
+
+	return
+}
+
+// Merges the vals' validator list with the updates list.
+// When two elements with same address are seen, the one from updates is selected.
+// Expects updates to be a list of updates sorted by address with no duplicates or errors,
+// must have been validated with verifyUpdates() and priorities computed with computeNewPriorities().
+func (vals *ValidatorSet) applyUpdates(updates []*Validator) {
+
+	existing := make([]*Validator, len(vals.Validators))
+	copy(existing, vals.Validators)
+
+	merged := make([]*Validator, len(existing)+len(updates))
+	i := 0
+
+	for len(existing) > 0 && len(updates) > 0 {
+		if bytes.Compare(existing[0].Address, updates[0].Address) < 0 {
+			merged[i] = existing[0]
+			existing = existing[1:]
+		} else {
+			merged[i] = updates[0]
+			if bytes.Equal(existing[0].Address, updates[0].Address) {
+				// validator present in both, advance existing
+				existing = existing[1:]
+			}
+			updates = updates[1:]
+		}
+		i++
+	}
+	for j := 0; j < len(existing); j++ {
+		merged[i] = existing[j]
+		i++
+	}
+	for j := 0; j < len(updates); j++ {
+		merged[i] = updates[j]
+		i++
+	}
+
+	vals.Validators = merged[:i]
+	vals.totalVotingPower = 0
```

Comment:
why reset `totalVotingPower`? Function declaration does not say anything about that

Reply:
This is probably leftover of old implementation. But we should probably avoid doing it and just always set it to a correct value.

Reply:
@melekes @milosevic ...yes I will look into it
valSet.TotalVotingPower() just returns valSet.totalVotingPower if it is non-zero. It should probably be enough to call valSet.TotalVotingPower() at the end of updateWithChangeSet()